presto-main/src/main/java/com/facebook/presto/catalogserver/RemoteMetadataManager.java

```diff
@@ -0,0 +1,780 @@
+import com.facebook.presto.spi.MaterializedViewStatus;
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.SystemTable;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.TableLayoutFilterCoverage;
+import com.facebook.presto.spi.connector.ConnectorCapabilities;
+import com.facebook.presto.spi.connector.ConnectorOutputMetadata;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.facebook.presto.spi.function.SqlFunction;
+import com.facebook.presto.spi.security.GrantInfo;
+import com.facebook.presto.spi.security.PrestoPrincipal;
+import com.facebook.presto.spi.security.Privilege;
+import com.facebook.presto.spi.security.RoleGrant;
+import com.facebook.presto.spi.statistics.ComputedStatistics;
+import com.facebook.presto.spi.statistics.TableStatistics;
+import com.facebook.presto.spi.statistics.TableStatisticsMetadata;
+import com.facebook.presto.sql.planner.PartitioningHandle;
+import com.facebook.presto.transaction.TransactionManager;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.slice.Slice;
+
+import javax.inject.Inject;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+// TODO : Use thrift to serialize metadata objects instead of json serde on catalog server in the future
+public class RemoteMetadataManager
+        implements Metadata
+{
+    // This value represents an empty result from the catalog server
+    private static final String EMPTY_STRING = "";
+
+    private final Metadata delegate;
+    private final TransactionManager transactionManager;
+    private final ObjectMapper objectMapper;
+    private final DriftClient<CatalogServerClient> catalogServerClient;
+
+    @Inject
+    public RemoteMetadataManager(
+            MetadataManager metadataManager,
+            TransactionManager transactionManager,
+            ObjectMapper objectMapper,
+            DriftClient<CatalogServerClient> catalogServerClient)
+    {
+        this.delegate = requireNonNull(metadataManager, "metadata is null");
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.objectMapper = requireNonNull(objectMapper, "objectMapper is null");
+        this.catalogServerClient = requireNonNull(catalogServerClient, "catalogServerClient is null");
+    }
+
+    @Override
+    public final void verifyComparableOrderableContract()
+    {
+        delegate.verifyComparableOrderableContract();
+    }
+
+    @Override
+    public Type getType(TypeSignature signature)
```

Comment:
I think we should have DelegatingMetadataManager which just delegate the implementation, and have this class implement the methods which it supports. This could be useful to implement testremotemanager or provide default stub.

Reply:
Doing this wouldn't we still need to implement all classes though? Since this implements from the Metadata interface we would need to provide an implementation for all of the methods as they dont provide any default methods.

Reply:
Implementing all classes is required, however with DelegatingMetadataManager we can have default implementation, and have RemoteMetadataManager implementing only required methods.

Reply:
I see, I added a separate commit for now doing that. Let me know what you think about it.